test/coinselector-test.js

```diff
@@ -0,0 +1,178 @@
+/* eslint-env mocha */
+/* eslint prefer-arrow-callback: "off" */
+
+'use strict';
+
+const assert = require('bsert');
+const {CoinSelector, UTXO} = require('../lib/wallet/coinselector');
+const TX = require('../lib/primitives/tx');
+
+function buildUTXOs(values) {
+  const utxos = [];
+  for (let i = 0; i < values.length; i++) {
+    // spending size of P2WPKH is 69
+    utxos.push(new UTXO(69, values[i], i));
+  }
+  return utxos;
+}
+
+const selector = new CoinSelector(new TX());
+const values = [100000, 50000, 30000, 20000, 10000, 5000, 3000, 2000, 1000];
+const utxos = buildUTXOs(values);
+selector.utxos = utxos;
+const costOfChange = 345; // this is cost of producing and spending a change output
+
+const targetSet1 = [221000, 220000, 215000, 214000, 211000, 208000, 206000, 203000, 201000,
+                    195000, 186000, 178000, 166000, 160000, 155000, 152000, 146000, 139000, 119000,
+                    116000, 110000, 109000, 108000, 106000, 105000, 101000, 98000, 96000, 90000,
+                    85000, 82000, 81000, 80000, 78000, 71000, 67000, 66000, 63000, 55000, 53000,
+                    51000, 45000, 44000, 41000, 38000, 36000, 23000, 19000, 16000, 11000, 6000];
+
+const targetSet2 = [150000, 130000, 101000, 50000, 15000, 13000, 5000, 3000];
+
+const targetSet3 = [219000, 217000, 213000, 212000, 211000, 205000, 202000, 201000, 190000,
+                    185000, 183000, 182000, 181000, 170000, 155000, 153000, 152000, 151000, 130000,
+                    120000, 110000, 105000, 103000, 102000, 101000];
+
+describe('Coin Selector', function() {
+  describe('Branch and Bound Selection', function() {
+    // try selecting single UTXOs
+    for (const value of values) {
+      it(`should select target=${value} using Branch and Bound`, () => {
+        const selection = selector.selectBnB(value, costOfChange);
+        assert.strictEqual(selection.length, 1);
+        assert.strictEqual(utxos[selection[0]].effectiveValue, value);
```

Comment:
In "test: added CoinSelector tests" (f181ffd3ea4f87d0364543804a0f76d9b8ef928c):
What does this test cover? Does it test whether BnB would select a single input of the exact matching value at a feerate of 0, or am I missing something?

Reply:
Yeah, that's correct. This is to check if BnB selects a single UTXO of the exact value and not other combinations.

Reply:
Perhaps you could add a comment to explain what you're testing for future readers of this code.